src/execution.rs

```diff
@@ -54,19 +56,31 @@ use crate::variables::Globals;
 use crate::variables::VariableMap;
 use crate::variables::Variables;
 use crate::Identifier;
+use crate::Location;
 
 impl File {
     /// Executes this graph DSL file against a source file.  You must provide the parsed syntax
     /// tree (`tree`) as well as the source text that it was parsed from (`source`).  You also
     /// provide the set of functions and global variables that are available during execution.
-    pub fn execute<'tree>(
+    pub fn execute<'a, 'tree>(
         &self,
         tree: &'tree Tree,
+        source_path: &'tree Path,
         source: &'tree str,
+        tsg_path: &'a Path,
+        tsg_source: &'a str,
```

Comment:
It _is_ possible to construct the `File` data structure directly and not from a source. However, one would also have to fake all the `location`s in there. Anyway, perhaps it is worth considering if anything panics if an empty source is passed here. (Panicking in error reporting seems something worth avoiding.)

Reply:
AFAICS no. There (I believe) only three new `unwrap`s, and the only unsafe one occurs when retrieving the root node for a match; if it panics, it's because one of our or `tree-sitter`'s invariants has failed.

The other two are when formatting `Excerpt`s, whose `source` is `Option`al (because what if there is no _n_th line of the indicated string?), and whose `path` is non-`Option`al but must be converted back to a `str` for display (which can fail). In both of these cases we unwrap with a default (`unwrap_or`).